pkg/controller/utils/component.go

```diff
@@ -278,6 +281,51 @@ func mergeState(desired, current runtime.Object) runtime.Object {
 	}
 }
 
+func ensureOSAffinity(obj runtime.Object, osTypes []render.OSType) {
+	if len(osTypes) == 0 {
+		return
+	}
+
+	var podTemplate *v1.PodTemplateSpec
+	switch obj.(type) {
+	case *apps.Deployment:
+		podTemplate = &obj.(*apps.Deployment).Spec.Template
+	case *apps.DaemonSet:
+		podTemplate = &obj.(*apps.DaemonSet).Spec.Template
+	case *apps.StatefulSet:
+		podTemplate = &obj.(*apps.StatefulSet).Spec.Template
+	case *batchv1beta.CronJob:
+		podTemplate = &obj.(*batchv1beta.CronJob).Spec.JobTemplate.Spec.Template
+	default:
+		return
+	}
+
+	// Use node selectors if there's only one supported OS type for backwards compatibility
+	if len(osTypes) == 1 {
+		if podTemplate.Spec.NodeSelector == nil {
+			podTemplate.Spec.NodeSelector = make(map[string]string)
+		}
+
+		podTemplate.Spec.NodeSelector["kubernetes.io/os"] = string(osTypes[0])
+	} else {
+		var values []string
+		for _, typ := range osTypes {
+			values = append(values, string(typ))
+		}
+
+		term := v1.NodeSelectorTerm{
+			MatchExpressions: []v1.NodeSelectorRequirement{{
+				Key:      "kubernetes.io/os",
+				Operator: v1.NodeSelectorOpIn,
+				Values:   values,
+			}},
+		}
+
+		rd := podTemplate.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution
```

Comment:
The thing is that if it's already scheduled and running then it would really matter as it's likely running on a supported node.

I'm 50 / 50 for changing it to be singular. Doing this now means that we don't need to update all the functions later, and if we never have multiple supported operating systems then it's not that big a deal, but on the other hand it really would be that difficult to update it later.

Reply:
Done